src/mono/System.Private.CoreLib/src/System/Threading/ObjectHeader.Mono.cs

```diff
@@ -0,0 +1,451 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Threading;
+
+/// <summary>
+/// Manipulates the object header located in the first few words of each object in the managed heap.
+/// </summary>
+internal static class ObjectHeader
+{
+    [StructLayout(LayoutKind.Sequential)]
+    private unsafe struct Header
+    {
+#region Keep in sync with src/native/public/mono/metadata/details/object-types.h
+        public void* vtable;
+        public IntPtr synchronization; // really a LockWord
+#endregion // keep in sync with src/native/public/mono/metadata/details/object-types.h
+    }
+
+    // This is similar to QCallHandler ObjectHandleOnStack, but with a getter that let's view the
+    // object's header.  This does two things:
+    //
+    // 1. It gives us a way to pass around a reference to the object header
+    //
+    // 2. because mono uses conservative stack scanning, we ensure there's always some place on the
+    // stack that stores a pointer to the object, thus pinning the object.
+    private unsafe ref struct ObjectHeaderOnStack
+    {
+        private Header** _header;
```

Comment:
If this is a Header** doesn't that mean there won't be a visible pointer to the object on the stack? Instead, it's a pointer-to-pointer, right?

Reply:
`_header` is a pointer to pointer.  Which means the thing that it points to is on the stack (well... the thing it points to is in memory - but we happen to know it's on the stack) - that's the one we want the GC to notice.